pkg/catalog/remote_catalog.go

```diff
@@ -130,48 +151,25 @@ func (c *RemoteCatalog) patchDiff(
 // as we already know the spec will change. It also saves us the use of any special status field.
 // If for some reason the generation is incremented multiple times in between the current and the next reconciliation
 // it can simply jump multiple generations by always basing it on the latest generation of the remote.
-func (c *RemoteCatalog) CalculateDiffs(
+func (c *RemoteCatalog) diffsToDelete(
 	runtimeList *v1alpha1.ModuleTemplateList, controlPlaneList *v1alpha1.ModuleTemplateList,
-) ([]*v1alpha1.ModuleTemplate, []*v1alpha1.ModuleTemplate) {
-	// these are various ModuleTemplate references which we will either have to create, update or delete from
-	// the remote
-	diffToApply := make([]*v1alpha1.ModuleTemplate, 0, len(runtimeList.Items))
-	var diffToDelete []*v1alpha1.ModuleTemplate
-
-	// now lets start using two frequency maps to discover diffs
-	existingOnRemote := make(map[string]int)
-	existingOnControlPlane := make(map[string]int)
-
-	for i := range runtimeList.Items {
-		remote := &runtimeList.Items[i]
-		existingOnRemote[remote.Namespace+remote.Name] = i
-	}
-
-	for i := range controlPlaneList.Items {
-		controlPlane := &controlPlaneList.Items[i]
-		existingOnControlPlane[controlPlane.Namespace+controlPlane.Name] = i
-		// if the controlPlane Template does not exist in the remote, we already know we need to create it
-		// in the runtime
-		if _, exists := existingOnRemote[controlPlane.Namespace+controlPlane.Name]; !exists {
-			c.prepareForSSA(controlPlane)
-			diffToApply = append(diffToApply, controlPlane)
+) []*v1alpha1.ModuleTemplate {
+	kcp := controlPlaneList.Items
+	skr := runtimeList.Items
+	toDelete := make([]*v1alpha1.ModuleTemplate, 0, len(runtimeList.Items))
+	for skrIndex := range skr {
```

Comment:
I'm fine with current implementation, but I think you can keep the existing way of calculating the `diffToDelete`, it should be fine, and on performance perspective, the previous one use hashmap is also better than current two for loop. And I think we should remove the previous function comments, it's still `CalculateDiffs` and the explanation is not up-to-date.

```
for i := range runtimeList.Items {
		remote := &runtimeList.Items[i]
		controlPlaneIndex, exists := existingOnControlPlane[remote.Namespace+remote.Name]

		// if the remote Template does not exist in the control plane, we already know we need to delete it
		if !exists {
			diffToDelete = append(diffToDelete, remote)
			continue
		}
	}
```

Reply:
I disagree on the performance I believe. The Hashmaps I do not need now since I am not iterating multiple times. Frequency Maps should only result in benefits If I plan to iterate multiple times over any of the 2 lists.

Regarding the comments, I fully agree, I will update these.

Reply:
I created a similar logic with frequency maps now but made it a bit more simple since we do not need to compute apply sets anymore (because we can simply always apply all templates, that way we don't have to compare them)